### PR TITLE
MCprep 3.5 Rewritten Core

### PR DESCRIPTION
**Pull request by StandingPadAnimations**
Alright, this is a new branch tracking deprecations and migrating 2.7x style code to 2.8. The term "Rewritten Core" I think acts as a nice theme for this migration, as it requires going into parts of MCprep that have long been forgotten about.

# Faq (for anyone that stumbles apon this)
 - Q: Will this have any performance improvements?
	 - A: No, any performance improvements will be negligible at most
 - Q: When's the earliest I can use this?
 	- A: As of writing, we can't say. When this does go into public testing (likely as 3.4.99), we'll create a GitHub discussion for bug reports and testing
 - Q: Why are you removing 2.7x support?
 	- A: See #399, but the gist is that the effort of maintaining compatibility is not worth it for the small amount of 2.7x users
- Q: Can I contribute code?
	- A: Sure, but you should have a good understanding of bpy and Git
- Q: I'm a 2.7x user, what does this mean for me?
	- A: This means you will have to either stick with the MCprep 3.4 series, or upgrade. I've talked with 2.7x users who've moved to modern versions of Blender, and according to them, as long as you choose the 2.7x keybinds, migrating is pretty painless

# Plans
- [ ] Migrate `PropertyGroups` to use annotations
- [x] Remove Blender Internal support 
- [ ] Add annotations to common functions
- [ ] Replace use of `os` with `pathlib` for file operations
- [ ] Start using dataclasses when it makes sense 
- [ ] Start using f-strings for readability

# Deprecations
## `util.make_annotations(cls)` and `util.bv28()`
Dropped as 2.7x support will also be dropped. Just use type annotations for properties instead of the old 2.7x syntax and don't target 2.7x when writing code

### CODE REVIEW

**Comment by StandingPadAnimations** [comment]
Btw @TheDuckCow, what errors did you get when testing? I can't run the test suite on my laptop (due to some OSX specific parts of the testing code)

**Comment by StandingPadAnimations** [comment]
I think for this branch I'll also start simplifying the code. For example, the following function:
```py
def matprep_cycles(
	mat, passes, use_reflections, use_principled, only_solid, pack_format, use_emission_nodes)
```

Could be adjusted as:
```py
@dataclass
class PrepOptions:
	passes: dict[str],
	use_reflections: bool,
	use_principled: bool,
	only_solid: bool,
	pack_format: str 
	use_emission_nodes: bool

def matprep_cycles(mat, options: PrepOptions)
```

**Comment by StandingPadAnimations** [comment]
> I think for this branch I'll also start simplifying the code. For example, the following function:
> 
> ```python
> def matprep_cycles(
> 	mat, passes, use_reflections, use_principled, only_solid, pack_format, use_emission_nodes)
> ```
> 
> Could be adjusted as:
> 
> ```python
> @dataclass
> class PrepOptions:
> 	passes: dict[str]
> 	use_reflections: bool
> 	use_principled: bool
> 	only_solid: bool
> 	pack_format: str 
> 	use_emission_nodes: bool
> 
> def matprep_cycles(mat, options: PrepOptions)
> ```

Alright I pushed thia change yesterday. @TheDuckCow, what do you think?

**Comment by TheDuckCow** [comment]
Hey @StandingPadAnimations I like that change, and am OK with it.

Once suggestion though: let's do changes a little differently:

Keep internal-rewrite as a "feature branch", treated in a similar was a dev (in fact, we'll merge dev into internal-rewrites on occasion to keep things in sync). Then, any new changes can be creates as their own branches, PR reviewed on their own, and merged into this internal branch. This way, we can both make incremental improvements without stumbling all in one mega PR. If you're aligned, I'd just do that for any further work on 3.5 moving forward.

**Comment by StandingPadAnimations** [comment]
Sure thing

**Comment by TheDuckCow** [comment]
Per [comment here](https://github.com/TheDuckCow/MCprep/pull/348), might be good to turn these into more incremental smaller branches merged into milestone-3-5-0 one at a time, instead of turning this to an ever-increasing in size PR. I'd consider capping the work done so far and see about merging that (but let me do another code review of what's done so far).

**Comment by StandingPadAnimations** [comment]
Yeah I think we should do that. It would also give developers the benefit of the new changes.

Just one problem: the addon updater is complaining a lot:
```sh
Traceback (most recent call last):
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 687, in draw
    addon_updater_ops.check_for_update_background()
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater_ops.py", line 816, in check_for_update_background
    updater.set_check_interval(enabled=settings.auto_check_update,
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater.py", line 558, in set_check_interval
    raise ValueError("Enable must be a boolean value")
ValueError: Enable must be a boolean value
Traceback (most recent call last):
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 1623, in draw
    addon_updater_ops.check_for_update_background()
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater_ops.py", line 816, in check_for_update_background
    updater.set_check_interval(enabled=settings.auto_check_update,
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater.py", line 558, in set_check_interval
    raise ValueError("Enable must be a boolean value")
ValueError: Enable must be a boolean value
Traceback (most recent call last):
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 687, in draw
    addon_updater_ops.check_for_update_background()
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater_ops.py", line 816, in check_for_update_background
    updater.set_check_interval(enabled=settings.auto_check_update,
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater.py", line 558, in set_check_interval
    raise ValueError("Enable must be a boolean value")
ValueError: Enable must be a boolean value
Traceback (most recent call last):
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 1623, in draw
    addon_updater_ops.check_for_update_background()
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater_ops.py", line 816, in check_for_update_background
    updater.set_check_interval(enabled=settings.auto_check_update,
  File "/home/mahid/.config/blender/3.5/scripts/addons/MCprep_addon/addon_updater.py", line 558, in set_check_interval
    raise ValueError("Enable must be a boolean value")
ValueError: Enable must be a boolean value
```

**Comment by TheDuckCow** [comment]
@StandingPadAnimations  Can you try first merging the base milestone branch into this to resolve issues? Then should be quick to resolve anything else like if something got messed up with updater preferences (I have no other reports spearately that there's anything wrong in blender 3.5 and the updater).

```
git checkout milestone-3-5-0
git pull
git checkout mcprep-3_5-recore
git merge milestone-3-5-0 
# and now resolve conflicts :D  personally a fan of sublime merge, but lots of
# tools out there, or command line. Just be sure to live test afterwards!
```

If you needed, I could also help out with this step. But the conflicts cover changes you made, so I feel you're a better judge of how to do the merge. Then I can swoop in to fix whatever with the updater here.

**Comment by StandingPadAnimations** [comment]
Just one problem: the world panel is not appearing D:

No errors in the console as far as I can tell, nor any UI changes that would suggest an issue. 

Anyway do you also want me to close the `internal-rewrites` PR since this branch also includes those changes?

**Review by TheDuckCow — CHANGES_REQUESTED** [review]
Given [this PR](https://github.com/TheDuckCow/MCprep/pull/418) is blocked on this now, @StandingPadAnimations we'll want to prioritize getting this ready and merged.

I've made a commit which adds what I believe are the remaining annotations, so we are fully moving to that and dropping the make_annotations workaround. Preferences initially wasn't working, now it is. However, something is still causing the `bpy.types.Scene.mcprep_props` to fail or otherwise not actually register, as I am getting UI draw errors saying:

```
Traceback (most recent call last):
  File "/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 908, in draw
    scn_props = context.scene.mcprep_props
AttributeError: 'Scene' object has no attribute 'mcprep_props'
```

I think this still may be annotation related - I recall errors regarding that were typically quite cryptic and misleading (I don't see any registration errors).

I'll keep debugging too and see what I can find, but feel free to try to beat me to it.

**Comment by TheDuckCow** [comment]
Alright, now we're in better shape. All UI seems to draw, though many of the unit tests are still failing, something we'll need to look into.  Some live testing didn't show anything terribly off, could just be test code needs updating.

@StandingPadAnimations let me know once you do some testing on this branch yourself (after doing a git pull), and see if it's looking good for you aside from these unit tests. If so, I'd say we can merge and then I can tackle the unit test fixing separately.

```
Failed tests:
	spawn_mob: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	spawn_mob_linked: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	change_skin: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	import_jmc2obj: "][mapping_set]\n TypeError: 'NoneType' object is not subscriptable\n 
	import_mineways_separated: "][mapping_set]\n TypeError: 'NoneType' object is not subscriptable\n 
	import_mineways_combined: "][mapping_set]\n TypeError: 'NoneType' object is not subscriptable\n 
	item_spawner: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	item_spawner_resize: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	geonode_effect_spawner: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	particle_area_effect_spawner: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	collection_effect_spawner: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	img_sequence_effect_spawner: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
	load_material: lender.app/Contents/Resources/3.5/scripts/modules/bpy/ops.py:113\n \n 
```

**Comment by TheDuckCow** [comment]
> Anyway do you also want me to close the `internal-rewrites` PR since this branch also includes those changes?

If that's indeed the case, those changes are fully here, then sure you can close that PR.

**Comment by StandingPadAnimations** [comment]
@TheDuckCow I've added a `pyproject.toml` file for Poetry (mainly to make my life easier when Python updates on Linux, that's an absolute nightmare). Let me know if anything needs to be changed.

It's just for the development environment, not runtime

**Comment by StandingPadAnimations** [comment]
> let me know once you do some testing on this branch yourself (after doing a git pull), and see if it's looking good for you aside from these unit tests.

Beyond the deprecation warnings regarding the use of `bv28()`, I think we're in good shape. I'll start working on those deprecation warnings

**Comment by StandingPadAnimations** [comment]
Ahh yes:
```py
factor_width = 0.3
if util.bv28():
    factor_width = 0.3
```

**Comment by StandingPadAnimations** [comment]
I've removed all deprecation warnings at startup and exit, which means now we can now start improving and testing the codebase (and live test to find more use of deprecated functions at the same time)

I think we can now perform the first merge with the milestone branch

**Comment by StandingPadAnimations** [comment]
Pushed some fixes for `compile.sh`, including 2 big errors caught by `shellcheck`. Read [`eccda95`](https://github.com/TheDuckCow/MCprep/pull/401/commits/eccda95b80ae0a6445015dd0d6743111992bb3da) for more info. 

Tldr: We were using some syntax that if done wrong could have done some serious damage to a user's system (on Linux)

`compile.sh` already has a rocky history with doing some pretty bad things (case in point, [this moment from January recorded on the MCprep server](https://discord.com/channels/737871405349339232/923762820242354246/1066884691778621490), which I finally fixed today), so we absolutely need to start linting shell scripts. https://github.com/koalaman/shellcheck is an amazing tool and would immediately start catching more errors. In addition, I think we need to start writing more cross platform scripts (I can't run the test suite due to OS X specific things in the script files)

**Comment by TheDuckCow** [comment]
@StandingPadAnimations 

> Pushed some fixes for `compile.sh`, including 2 big errors caught by `shellcheck`. Read [`eccda95`](https://github.com/TheDuckCow/MCprep/pull/401/commits/eccda95b80ae0a6445015dd0d6743111992bb3da) for more info.
> 
> Tldr: We were using some syntax that if done wrong could have done some serious damage to a user's system (on Linux)

Great to have these fixes - but let's try to not expand the scope of this PR any more than it already has. It would be better to merge this into `milestone-3-5-0` _as soon as we are ready with no further changes_ (outside of _fixes_ to what this branch already changes) as soon as possible.

Even changes like the `pyproject.toml` I want to think about more, but it's now wrapped up in all these other changes here (I'm ok with going with it though since it's dev side only, it's just an example). A good goal moving forward is small PRs (ie like 100 lines of code) that address one explainable net change (composed of one or more commits that make that single change a reality), and then you can have one PR chained on top of another if there's dependencies of one change on a prior and you don't want to wait for the first PR to be merged. 

So, let's do no more new changes on this branch. I need to spot check the recent changes/additions to see if anything needs addressing, and then based on my approval, let's merge, and create new branches PRs off of newly updated `milestone-3-5-0`.

File: MCprep_addon/mcprep_ui.py [hunk 1]
```diff
@@ -915,7 +893,7 @@ class MCPREP_PT_world_tools(bpy.types.Panel):
 	"""World settings and tools"""
 	bl_label = "World Tools"
 	bl_space_type = 'VIEW_3D'
-	bl_region_type = 'TOOLS' if not util.bv28() else 'UI'
+	bl_region_type = 'UI'
```
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:896]: Nit on extra space at end of line. Not that there aren't more egregious code style issues pre-existing 😆

File: MCprep_addon/mcprep_ui.py [hunk 2]
```diff
@@ -1713,7 +1688,7 @@ class MCPREP_PT_mob_spawner(bpy.types.Panel):
 	bl_label = "Mob spawner"
 	bl_parent_id = "MCPREP_PT_spawn"
 	bl_space_type = "VIEW_3D"
-	bl_region_type = 'TOOLS' if not util.bv28() else 'UI'
+	bl_region_type = 'UI'
```
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:1691]: (same here)

File: MCprep_addon/mcprep_ui.py [hunk 3]
```diff
@@ -2093,13 +2068,9 @@ def register():
 		default=addon_prefs.custom_texturepack_path)
 
 	env.verbose = addon_prefs.verbose
-	if hasattr(bpy.types, "INFO_MT_add"):  # 2.7
-		bpy.types.INFO_MT_add.append(draw_mcprepadd)
-	elif hasattr(bpy.types, "VIEW3D_MT_add"):  # 2.8
+	if hasattr(bpy.types, "VIEW3D_MT_add"):  # 2.8
```
Comment (TheDuckCow) [inline MCprep_addon/mcprep_ui.py:2071]: Here and below, might be god to have a printout in an else clause - that way there still isn't a registration error, but we'd be aware of it when looking at startup logs.

Also in unregister, a few remaining blocks which can be simplified to remove the 2.7 logic branches. Same deal to just have an else clause that prints a warning.

File: poetry.lock [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# This file is automatically @generated by Poetry 1.4.2 and should not be changed by hand.
```
Comment (TheDuckCow) [inline poetry.lock:1]: What's the consequence of this file if, for example, you were to step away and I didn't update this with future updates? Will this file need to change with releases, or is it expected to remain mostly static? Would these hashes be universal, or system specific to you?

**Review by TheDuckCow — APPROVED** [review]
Largely approving, left a couple comments. Still feel the bash + poetry could be better suited in their own branches, but for simplicity, I'm ok if we have it here and just going forward make dedicated PRs.

File: poetry.lock (hunk 4 above)
Comment (StandingPadAnimations) [reply poetry.lock:1]: According to the poetry documentation, the hashes are based on the dependencies. It'd automatically handled by poetry

File: MCprep_addon/mcprep_ui.py (hunk 1 above)
Comment (StandingPadAnimations) [reply MCprep_addon/mcprep_ui.py:896]: I'm considering adding https://github.com/akaihola/darker in a later PR so that we can gradually add better formatting, so hopefully this will be resolved in the future

**Comment by StandingPadAnimations** [comment]
Ight, hopefully I'll merge this later today
